File: Algorithms/169 - Majority Element/01 - Majority Element.cpp

  // 这种解法是使用一个hash表map<>，键用来存放数组的元素，键对应的值存放元素出现的次数。遍历整个数组，查找它在hash表中是否出现，如果出现将出现次数加1，如果没有出现，将它插入hash表中，并设置它的出现次数为1。每次遍历到一个元素，判断它的出现次数是否超过了数组长度的一半，要是超过了就返回该元素。时间复杂度是O(n)，空间复杂度是O(n)。

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        if (nums.size() == 1)
            return nums[0];
        map<int,int> tables;
        for (int i = 0; i < nums.size(); i ++) {
            if (tables.count(nums[i])) {
                tables[nums[i]] ++;
                if (tables[nums[i]] > nums.size()/2)
                    return nums[i];
            }else {
                tables[nums[i]] = 1;
            }
        }
    }
};

File: Algorithms/169 - majorityElement（多数元素）/01.majorityElement.js
/**
 * 169. 多数元素
 * 
 * 给定一个大小为 n 的数组 nums ，返回其中的多数元素。多数元素是指在数组中出现次数 大于 ⌊ n/2 ⌋ 的元素。

你可以假设数组是非空的，并且给定的数组总是存在多数元素。

 

示例 1：

输入：nums = [3,2,3]
输出：3
示例 2：

输入：nums = [2,2,1,1,1,2,2]
输出：2
 
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function(nums) {
  let res = nums[0], count = 0;
  for (const num of nums) {
    if (count === 0) {
      res = num;
    }
    (num === res) ? count ++ : count --;
  }
  return res;
};